### ISSUE TITLE
Lots of CPU seemingly used for no reason.

### ISSUE THREAD

**Issue by addshore** [DEVSENSE/phptools-docs#99]
**Reproduction:**
 - Have a large PHP project
 - Open a file
 - Add some whitespace (or make any other small change?)
 - Save the file
 - CPU for phptools stays very elevated for a very long time

**Example:**
I added the newline highlighted in green in the image below.
For a period of 1.5 minutes my CPU was at 50% or more across 4 cores.
![image](https://user-images.githubusercontent.com/3308769/98650484-de4e2e00-2330-11eb-8529-60acc13e8c46.png)

**Expected behaviour:**
phptools would only do complex processing when actually needed, and thus not slow down machines for no reason.

**CPU screenshots example:**
![image](https://user-images.githubusercontent.com/3308769/98650209-7861a680-2330-11eb-94db-012910329ac6.png)
![image](https://user-images.githubusercontent.com/3308769/98650236-83b4d200-2330-11eb-8d29-9f5ea114d38d.png)

**Comment by jakubmisek** [comment]
Thanks @addshore , that's a lot. We'll try to find out more and to avoid similar CPU spikes.

In case you haven't tried it yet, you can disable workspace-wide analysis if it helps:

setting `"php.problems.workspaceAnalysis": false`
as described at https://docs.devsense.com/en/vscode/configuration#configuration-options

**Comment by addshore** [comment]
I think workspace wide analysis is very important.
But I don't think it should be doing something as cpu intensive as workspace wide analysis after every change.

**Comment by jakubmisek** [comment]
@addshore you are right, but it is designed to run for a few seconds, certainly not more than 5. We have to find out what's causing the CPU usage here.

**Comment by jakubmisek** [comment]
Does the CPU spikes only when you save the file? or it happens when you make a change (wait for 3 secs) without saving the file?

**Comment by addshore** [comment]
> Does the CPU spikes only when you save the file? or it happens when you make a change (wait for 3 secs) without saving the file?

I have autosave turned on, so by default I can't really make it only change and not save a file.
If you want me to investigate further though by turning autosave off I could have a look?

**Comment by jakubmisek** [comment]
We'll keep testing, also the memory usage of 4GB is very high. The most we achieved was 1GB (but on Windows)

How many `.php` files are there in your workspace please?

**Comment by addshore** [comment]
16709 PHP files.

```
---------------------------------------------------------------------------------------
Language                             files          blank        comment           code
---------------------------------------------------------------------------------------
PHP                                  16709         330715         721345        2305880
```

Full output here as there are some other things like html that I guess devsense might look at?
https://phabricator.wikimedia.org/P13632

**Comment by addshore** [comment]
> The most we achieved was 1GB (but on Windows)

So one difference for me personally is that I am using VSCode in "remote" mode with WSL2.
So not pure windows, and this will cause some small overhead probably.

**Comment by addshore** [comment]
I can confirm that this seemingly random CPU usage still seems to be happening.
I actually just closed all of the files in my editor and then this happened.

![image](https://user-images.githubusercontent.com/3308769/103276913-ab82f680-49bf-11eb-895d-24ea6de9ba29.png)
This time only lasting for ~15-20 seconds.

**Comment by addshore** [comment]
Is there any way to see super verbose / trace logging of what is going on in the php lang server?

**Comment by jakubmisek** [comment]
In recent updates, we've optimized the process (still it takes about 1-3 seconds in our test environments). A trace is a good idea for a next update.

**Comment by jakubmisek** [comment]
the update `1.2.5783` optimizes the CPU usage and reduces the number of false problems reported.

**Comment by piniol** [comment]
Hi, on Ubuntu 20.04 after last updates CPU usage jumps to 100% after vscode is closed. Devsense.PHP.LanguageServer is not closing with vscode, just sits in process list and eats up cpu...

**Comment by piniol** [comment]
In addition: cpu usage is normal while vscode is running.

**Comment by jakubmisek** [comment]
@piniol thank you for the feedback - we have just noticed this as well on ubuntu and macOs. Working on it.

**Comment by jakubmisek** [comment]
@piniol the issue has been fixed and updated - since the version `1.2.5973` (Feb 08 2021)

I'll close the issue, please start a new one if the problem persists.
